test/e2e/apimachinery/watch.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apimachinery
+
+import (
+	"time"
+
+	"k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	watchPodLabelKey    = "watch-this-pod"
+	watchPodLabelValueA = "AAA"
+	watchPodLabelValueB = "BBB"
+)
+
+var _ = SIGDescribe("Watchers", func() {
+	f := framework.NewDefaultFramework("watch")
+
+	It("should observe add, update, and delete events on pods", func() {
+		c := f.ClientSet
+		ns := f.Namespace.Name
+
+		By("creating multiple similar watches on pods")
+		watchA, err := watchPodsWithLabels(f, watchPodLabelValueA)
+		Expect(err).NotTo(HaveOccurred())
+
+		watchB, err := watchPodsWithLabels(f, watchPodLabelValueB)
+		Expect(err).NotTo(HaveOccurred())
+
+		watchAB, err := watchPodsWithLabels(f, watchPodLabelValueA, watchPodLabelValueB)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Wait 10s for the watches to take effect
+		time.Sleep(10 * time.Second)
```

Comment:
Are there better way to do this (e.g. polling and check watcher's status)?

Reply:
Actually I think the .Watch() cakk won't return until the watch has taken effect. The test still passes without this line.